processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,37 +133,57 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
+    Integer priorIdex = facts.get(key);
+    final Boolean hadPriorValue = null != priorIdex;
+
+    Aggregator[] aggs = hadPriorValue ? aggList.get(priorIdex) : new Aggregator[metrics.length];
+
+    if (!hadPriorValue) {
```

Comment:
can we just have a simple if (prev != null) { ... } else { ... }, settings the value aggs inside of the if statement?